realm/realm-library/src/objectServer/java/io/realm/mongodb/sync/RecoverOrDiscardUnsyncedChangesStrategy.java

```diff
@@ -0,0 +1,83 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.realm.mongodb.sync;
+
+import io.realm.Realm;
+
+/**
+ * Strategy that tries to automatically recover any unsynced changes during a Client Reset.
+ * <p>
+ * A synced Realm may need to be reset because the MongoDB Realm Server encountered an error and had
+ * to be restored from a backup or because it has been too long since the client connected to the
+ * server so the server has rotated the logs.
+ * <p>
+ * The Client Reset thus occurs because the server does not have the full information required to
+ * bring the Client fully up to date.
+ * <p>
+ * In the event that the client reset cannot be automatically recover the unsynced data, it would
+ * try to discard any unsynced changes.
+ * <p>
+ * The discard unsynced changes reset process is as follows: when a client reset is triggered
+ * the {@link #onBeforeReset(Realm)} callback is invoked, providing an instance of the
+ * Realm before the reset and another after the reset, both read-only. Once the recover has concluded
+ * successfully the callback {@link #onAfterRecovery(Realm, Realm)} would be invoked, if it had to 
+ * discard the changes {@link #onAfterDiscard(Realm, Realm)} (Realm, Realm)} would be called.
+ * <p>
+ * If discarding the unsynced data is not enough to resolve the reset the
+ * {@link #onError(SyncSession, ClientResetRequiredError)} would be invoked, it allows to manually
+ * resolve the reset as it would be done in
+ * {@link ManuallyRecoverUnsyncedChangesStrategy#onClientReset(SyncSession, ClientResetRequiredError)}.
+ */
+public interface RecoverOrDiscardUnsyncedChangesStrategy extends SyncClientResetStrategy {
+
+    /**
+     * Callback that indicates a Client Reset is about to happen. It provides with a frozen instance
+     * of the Realm that is will be reset.
+     *
+     * @param realm frozen {@link Realm} in its state before the reset.
+     *
+     */
+    void onBeforeReset(Realm realm);
+
+    /**
+     * Callback invoked once the Client Reset has recovered the unsynced changes successfully.
+     * It provides of two Realm instances, a frozen one displaying the state before the reset and a
+     * regular Realm with the current state.
+     *
+     * @param before {@link Realm} frozen Realm in the before after the reset.
+     * @param after  {@link Realm} Realm after the reset.
+     */
+    void onAfterRecovery(Realm before, Realm after);
+
+    /**
+     * Callback invoked once the Client Reset has discarded the unsynced changes because it couldn't
+     * recover them. It provides of two Realm instances, a frozen one displaying the state before the
+     * reset and a regular Realm with the current state, that can be used to recover objects from the reset.
+     *
+     * @param before {@link Realm} frozen Realm in the before after the reset.
+     * @param after  {@link Realm} Realm after the reset.
```

Comment:
```suggestion
     * Callback invoked once the Client Reset has discarded the unsynced changes because it couldn't
     * recover them. It provides two Realm instances, a frozen one displaying the state before the
     * reset and a regular Realm displaying the current state that can be used to recover objects from the reset.
     *
     * @param before {@link Realm} frozen Realm in the state before the reset.
     * @param after  {@link Realm} Realm after the reset.
```